gh-128384: Use a context variable for warnings.catch_warnings

nascheme opened the pull request:
Make `warnings.catch_warnings()` use a context variable for holding the warning filtering state if the `sys.flags.context_aware_warnings` flag is set to true.  This makes using the context manager thread-safe in multi-threaded programs and safe for asyncio coroutines and tasks.  The flag is true by default in free-threaded builds and is otherwise false.  The value of the flag can be overridden by the the `-X thread_safe_warnings` command-line option or by the `PYTHON_CONTEXT_AWARE_WARNINGS` environment variable.

It is expected that one day the flag might default to true for all Python builds, not just the free-threaded one.  However, I think it makes sense to not commit to a schedule to do that until we have a better idea of what code is affected by this change.  Feedback from people using the free-threaded build should provide guidance.

This PR is on top of gh-128209.

A previous version of this PR used a single flag to control both behavior of `Thread` inheriting the context and also the `catch_warnings` context manager.   However, based on some feedback, I've decided that two flags makes things more clear.  Likely programs would like to set both flags to true to get the most intuitive behavior.

When the `context_aware_warnings` flag is true and a `catch_warnings()` context is active, added filters go into a list of filters stored in the contextvar, not the `warnings.filters` list.  The filters in `warnings.filters` are still applied but they apply after the contextvar filters.

That difference with how `warnings.filters` works is probably the most likely thing to cause broken user code.  In the unit tests, I had to change a number of references to `warnings.filters` to `warnings._get_filters()`.  That function returns the list of filters from the current context or the global filters if there is no context active.  Perhaps `_get_filters()` should become a public function?  I think it would be better if examining and manipulating that list was discouraged and so that's why I made the function non-public, at least for now.

I created `_py_warnings.py` to contain the Python implementation for the warnings module.  This matches other modules like `decimal` that have Python and C versions of the same module.   In the case of `warnings` things are a bit more complicated since you can assign to module globals (the `filters` list or the `showwarning()` function are common to override).  This is a cleaner organization, IMHO.  Now `warnings.py` just imports the parts it needs and the unit tests are a bit simpler.

I cleaned up the unit tests for `warnings` a bit.  Passing `module` to `catch_warnings()` actually serves no purpose.  Since `sys.modules['warnings']`  is already swapped to the module under test, the code does the right thing even when that parameter is not given.  I also replaced calls to `original_warnings.catch_warnings()` with `self.module.catch_warnings()`.  Both those calls do the same thing but the second spelling seems less confusing to me.

<!-- gh-issue-number: gh-128384 -->
* Issue: gh-128384
<!-- /gh-issue-number -->

[pyperformance comparison](https://github.com/facebookexperimental/free-threading-benchmarking/blob/main/results/bm-20250219-3.14.0a5%2B-dba89e0-NOGIL/bm-20250219-vultr-x86_64-nascheme-gh_128384_warnings_c-3.14.0a5%2B-dba89e0-vs-base.svg)

<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--130010.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

nascheme committed:
Add 'context' parameter to Thread.

* Add ``sys.flags.inherit_context``.
* Add ``-X inherit_context`` and :envvar:`PYTHON_INHERIT_CONTEXT`

nascheme committed:
Use contextvar for catch_warnings().

nascheme committed:
Make _contextvars a builtin module.

nascheme committed:
Tweak blurb markup.

nascheme committed:
Doc markup fix.

nascheme committed:
Add blurb.

nascheme committed:
Add "_warnings_context" as identifier.

nascheme committed:
Fix test_support for context var filters.

Since the unittest/runner.py establishes a new warnings context with
`catch_warnings`, these tests must use `warnings._get_filters()` to
retrieve the current list of filters, not the `warnings.filters` global.

nascheme committed:
Regenerate 'configure' script.

nascheme committed:
Regenerate 'configure' script.

nascheme committed:
Rename flag to `thread_inherit_context`.

nascheme committed:
Merge branch 'main' into thread_inherit_context

nascheme committed:
Merge branch 'thread_inherit_context' into gh-128384-warnings-contextvar

nascheme committed:
Use separate flag `thread_safe_warnings`.

nascheme committed:
Add doc for ``thread_safe_warnings`` flag.

kumaraditya303 commented on Python/_warnings.c:
```diff
@@ -254,6 +254,87 @@ warnings_lock_held(WarningsState *st)
+static PyObject *
+get_warnings_context(PyInterpreterState *interp)
+{
+    PyObject *ctx_var = GET_WARNINGS_ATTR(interp, _warnings_context, 0);
```
How about creating `_warnings_context` contextvar as a state variable in C implementation then override the pure python implementation with it? That would make working in C implementation much easier.

nascheme replied:
That's a good idea.  So much so that I already tried to do that.  :-)   I ran into difficulties making the tests work.  They import the modules two times in order to test both the Python and C versions of the module.  I'm not a fan of how those tests are organized so maybe I should just re-write a bunch of things there.  I was trying to avoid making too many changes.

Making `_warnings_context` a variable inside the C extension would be nicer.  We can't do that with `warnings.filters` because user code assigns it.  But we could with `_warnings_context`.  It can be done and it's cleaner (probably better performing too) and I'll try again.

nascheme replied:
It took some extra modifications to `warnings.py` but I got this to work.  `_warnings` now has a `_warnings_context` attribute and that's used by the C extension, rather than having to look into `sys.modules['warnings]` to get it.

nascheme committed:
Create _py_warnings.py for Python implementation.

nascheme committed:
Add _warnings_context to _warnings module.

nascheme committed:
Remove '_warnings_context' global string.

nascheme committed:
Rename flag to 'context_aware_warnings'.

nascheme committed:
Use catch_warnings() for module under test.

This actually doesn't make a difference for behaviour because everything
uses sys.modules['warnings'] anyhow (which is swapped out) but it's less
confusing this way.

nascheme committed:
Don't pass 'module' to catch_warnings().

This doesn't actually do anything given that sys.modules['warnings'] has
already been swapped out.  Simplify the unit test code by not passing
it.

nascheme committed:
Correct error in warnings module docs.

gpshead commented on Lib/_py_warnings.py:
```diff
@@ -1,37 +1,155 @@
+# of the 'warnings' module, commonly 'filters' or 'showwarning'. So we
+# need to lookup these global attributes dynamically on the '_wn' object,
+# rather than binding them earlier.  The code in this module consistently uses
+# '_wn.<something>' rather than using the globals of this module.  If the
```
```suggestion
# '_wm.<something>' rather than using the globals of this module.  If the
```

gpshead commented on Lib/_py_warnings.py:
```diff
@@ -1,37 +1,155 @@
+    def copy(self):
+        context = _Context(self._filters[:])
+        if self.log is not None:
+            context.log = self.log
```
this assigns a reference to the same list instead of a copy.  desired?  the if above suggests not.  add comments here if this isn't supposed to be the obvious copy operation of `context.log = self.log[:]` or if the above is not None condition is somehow meaningful even though this would just replace None with None as written.

nascheme replied:
I think not copying `.log` is the behavior that we desire.  If you have the following code:
```
with warnings.catch_warnings(record=True) as log:
   with warnings.catch_warnings("default", ...):
       ...
```
Then the logging should still be active inside the second `with` and it should log into the same list.  This matches what happens if you use the context handler with `_use_context = False`.

gpshead commented on Lib/test/test_decimal.py:
```diff
@@ -1725,8 +1726,10 @@ def test_threading(self):
 
-        th1 = threading.Thread(target=thfunc1, args=(self,))
-        th2 = threading.Thread(target=thfunc2, args=(self,))
+        th1 = threading.Thread(target=thfunc1, args=(self,),
```
add a comment as to why these need an explicit new context.

This update feels like the kind of example that'll be needed in a "porting code to be free-threading compatible" document for users of decimal and threads.  do we have a place in our docs to collect things like that?

nascheme replied:
I've added comment to clarify what's happening in that test.

AFAIK, we don't have a place to document such things.  In the case of the decimal module, if you want all threads to use the same decimal context, it's suggested you modify `decimal.DefaultContext`.  That will still work.  If you are using the default build and you want threads to inherit the context, I think you should just set the `sys.flags.thread_inherit_context` flag.  If you are some library code and don't get to control that, I think you need to assume new threads don't inherit (you don't know of the value of the flag being used) and ensure new threads set their initial context in some other way (like the `DefaultContext` mechanism used by decimal).

For `warnings.catch_warnings()` it's not as easy.  If `flags.context_aware_warnings` is false then it's just not safe to use that context manager from concurrent code.   If you care, again I think setting that flag would be the best fix.  Otherwise you have to do something like [Jax ](https://github.com/jax-ml/jax/pull/25626) does and write your own context manager for controlling warnings.  I think it's quite bad if we force a bunch of 3rd party libraries to write their own context managers and so I'm keen to get `warnings` fixed in 3.14.

gpshead commented on Lib/_py_warnings.py:
```diff
@@ -1,37 +1,155 @@
+
+
+def _get_filters():
+    """Return the current list of filters.  This is a non-public API used by
```
But _why_ do the unittests now all use a `mod._get_filters()` call instead of the more obvious pattern of just accessing `mod.filters` like end user code does?

nascheme replied:
The unit tests want to get the list of currently active filters, whether they are stored in `warnings.filters` (if `_use_context = False` or no `catch_warnings()` context is active) or in the context var.

I can't think of a reason why something other than the unit tests need to get the list of current filters.  However, maybe there is some reason and then it should be `get_filters()` and be a documented function.  My inclination is to keep things as a private functions unless there seems to be a real need for them to be used externally to the CPython codebase.

gpshead reviewed (commented):
I'm

kumaraditya303 commented on Python/_warnings.c:
```diff
@@ -254,6 +262,69 @@ warnings_lock_held(WarningsState *st)
+        Py_DECREF(ctx);
+        Py_RETURN_NONE;
+    }
+    PyObject *context_filters = PyObject_GetAttrString(ctx, "_filters");
```
Using static strings would be more performant
```suggestion
    PyObject *context_filters = PyObject_GetAttr(ctx, &_Py_ID(_filters));
```

kumaraditya303 commented on Python/_warnings.c:
```diff
@@ -254,6 +262,69 @@ warnings_lock_held(WarningsState *st)
+    }
+    if (ctx == Py_None) {
+        Py_DECREF(ctx);
+        Py_RETURN_NONE;
```
```suggestion
        return ctx;
```

kumaraditya303 commented on Python/_warnings.c:
```diff
@@ -401,42 +451,96 @@ get_filter(PyInterpreterState *interp, PyObject *category,
+    }
+    if (context_filters == Py_None) {
+        use_global_filters = true;
+        Py_DECREF(context_filters);
```
No need for decref, `None` is immortal.

nascheme committed:
Use PyObject_GetAttr.

nascheme committed:
Typo fix for docstring.

nascheme committed:
Avoid DECREF calls on None.

nascheme committed:
Add warnings.py module (missed in previous commit).

nascheme committed:
Add comment about why 'context' is passed in test.

nascheme committed:
Revise "decimal' docs, adding note about flag.

nascheme committed:
Merge branch 'thread_inherit_context' into gh-128384-warnings-contextvar

kumaraditya303 commented on Python/_warnings.c:
```diff
@@ -347,35 +417,14 @@ get_default_action(PyInterpreterState *interp)
+              PyObject *text, Py_ssize_t lineno,
+              PyObject *module, char *list_name, PyObject *filters,
+              PyObject **item, PyObject **matched_action) {
+    /* filters list could change while we are iterating over it. */
```
You would need critical section around this list otherwise concurrent mutations would crash in free-threading or switch to using strong ref like `PyList_GetItemRef`

nascheme committed:
Fix race in filter_search().

Using the critical section seems the easiest fix for this.  Add a
unit test that fails TSAN check if the fix is not applied.

nascheme committed:
Add note to free-threading howto.

nascheme committed:
Merge branch 'main' into gh-128384-warnings-contextvar

nascheme committed:
Doc fixes for missing refs.

nascheme committed:
Add "_py_warnings" to stdlib_module_names.h.

kumaraditya303 commented on Python/_warnings.c:
```diff
@@ -254,6 +262,68 @@ warnings_lock_held(WarningsState *st)
+    }
+    if (!PyList_Check(context_filters)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "warnings._warnings_context _filters must be a list");
```
```suggestion
                        "_filters of warnings._warnings_context must be a list");
```

kumaraditya303 commented on Python/_warnings.c:
```diff
@@ -401,42 +454,101 @@ get_filter(PyInterpreterState *interp, PyObject *category,
+           PyObject *text, Py_ssize_t lineno,
+           PyObject *module, PyObject **item)
+{
+    WarningsState *st = warnings_get_state(interp);
```
This needs to be wrapped in Py_DEBUG ifdef

kumaraditya303 commented on Python/_warnings.c:
```diff
@@ -401,42 +454,101 @@ get_filter(PyInterpreterState *interp, PyObject *category,
+    bool use_global_filters = false;
+    if (context_filters == NULL) {
+        return NULL;
+    }
```
```suggestion
    if (context_filters == NULL) {
        return NULL;
    }
    bool use_global_filters = false;
```

nascheme committed:
Improve error text in Python/_warnings.c

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

nascheme committed:
Avoid unused-variable warning in _warnings.c.

nascheme committed:
Minor code improvement to _warnings.c.

nascheme committed:
Merge 'origin/main' into gh-128384-warnings-contextvar

nascheme committed:
Merge branch 'origin/main' into gh-128384-warnings-contextvar

nascheme committed:
Merge 'origin/main' into gh-128384-warnings-contextvar

nascheme committed:
Add extra unit tests.

Check behaviour of the catch_warnings() context manager when used with
asyncio co-routines and threads.

nascheme committed:
Use asyncio events to get deterministic execution.

kumaraditya303 commented on Doc/howto/free-threading-python.rst:
```diff
@@ -152,3 +152,33 @@ to re-enable it in a thread-safe way in the 3.14 release.  This overhead is
+:class:`threading.Thread` to start with a copy of the
+:class:`~contextvars.Context()` of the caller of
+:meth:`~threading.Thread.start`.  If the flag is false, threads start with an
+empty :class:`~contextvars.Context()`.
```
```suggestion
In the free-threaded build, the flag :data:`~sys.flags.thread_inherit_context`
is set to true by default which causes threads created with
:class:`threading.Thread` to start with a copy of the
:class:`~contextvars.Context()` of the caller of
:meth:`~threading.Thread.start`.  In the default GIL-enabled build, the flag
defaults to false so threads start with an
empty :class:`~contextvars.Context()`.
```

nascheme committed:
Update Doc/howto/free-threading-python.rst

Co-authored-by: Kumar Aditya <kumaraditya@python.org>

nascheme committed:
Merge 'origin/main' into gh-128384-warnings-contextvar

nascheme committed:
Add unit test for asyncio tasks.

nascheme committed:
Use asyncio.gather() rather than create_task().

This seems a bit simpler since gather() will create tasks as needed.

nascheme committed:
Call resetwarnings() in a few places.

This is needed because unit tests are now run with "-W error" on the
command line in CI (GH-128770).  We want to start with an empty list of
filters to avoid supurious errors.
